added autofocus in search modal

alexrybch opened the pull request:
Signed-off-by: Alex Rybchenko <arybchenko@box.com>

## Hey, I just made a Pull Request!

Added autofocus for search modal
Added redicrect to full search results on "Enter" press

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

emmaindal commented:
Hey @alexrybch thanks for contribution, just some initial thinking

Because of the fact that auto focus for input fields is not always accessible, we made this optional for the adopter. Thats why we added the possibility of using a search bar that is focused https://backstage.io/storybook/?path=/story/plugins-search-searchbar--focused. With that said, this PR is updating the SearchModal to always use a searchbar which is focused, and that doesn't give the adopter possibility to choose. I wonder if we should instead refactor the search modal so that it becomes a bit more composable (similar to the [SearchPage](https://backstage.io/docs/features/search/getting-started#adding-search-to-the-frontend)). Something like: 

```tsx
<SearchModal open={open} toggleModal={toggleModal}>
    <SearchBar autoFocus />
    ....
    results etc
</SearchModal>
```

that will gives us the opportunity to still let adopters decide if they want the focused search bar or not. What do you think about this?

alexrybch commented:
> Hey @alexrybch thanks for contribution, just some initial thinking
> 
> Because of the fact that auto focus for input fields is not always accessible, we made this optional for the adopter. Thats why we added the possibility of using a search bar that is focused https://backstage.io/storybook/?path=/story/plugins-search-searchbar--focused. With that said, this PR is updating the SearchModal to always use a searchbar which is focused, and that doesn't give the adopter possibility to choose. I wonder if we should instead refactor the search modal so that it becomes a bit more composable (similar to the [SearchPage](https://backstage.io/docs/features/search/getting-started#adding-search-to-the-frontend)). Something like:
> 
> ```tsx
> <SearchModal open={open} toggleModal={toggleModal}>
>     <SearchBar autoFocus />
>     ....
>     results etc
> </SearchModal>
> ```
> 
> that will gives us the opportunity to still let adopters decide if they want the focused search bar or not. What do you think about this?

good point! I added optional searchBar props to make it more composable, so as a bonus other props like `placeholder` can be overriden too

camilaibs commented on plugins/search/src/components/SearchModal/SearchModal.tsx:
```diff
@@ -41,6 +41,10 @@ import { useNavigateToQuery } from '../util';
   toggleModal: () => void;
+  searchBarProps?: Omit<
+    SearchBarBaseProps,
+    'value' | 'className' | 'onChange' | 'onSubmit'
```
Is there a reason to omit the `className`?

alexrybch replied:
There's already `classes.input` passed as className for SearchBar. Do you think `classnames` should be used and allow to extend className ?

freben replied:
Could be that this should be unified - i.e. that people should move over to passing in `SearchBarBaseProps.className` instead of `classes.input`? But maybe in a separate PR if so, what you do here is probably OK for the current state of the code around your change

camilaibs commented on plugins/search/src/components/SearchModal/SearchModal.tsx:
```diff
@@ -99,10 +114,11 @@ export const Modal = ({ open = true, toggleModal }: SearchModalProps) => {
       <DialogTitle>
         <Paper className={classes.container}>
           <SearchBarBase
+            {...searchBarProps}
```
There is a problem asking whether we should avoid multiple search bar implementations here.
Do you know the context of why we are using `<SearchBarBase />` instead of `<SearchBar />` @emmaindal ?
The `SearchBar` already has a `debounceTime` property ready to use, for example. I would like to know if we can replace here making the `SearchBar` props customizable as the Material UI `TextField` is...

camilaibs reviewed (approved):
Thanks again, small comments only!

freben commented on plugins/search/api-report.md:
```diff
@@ -158,6 +159,13 @@ export const SearchModal: ({
 export interface SearchModalProps {
   // (undocumented)
   open?: boolean;
+  // Warning: (ae-forgotten-export) The symbol "SearchBarBaseProps" needs to be exported by the entry point index.d.ts
```
Before merging, I think we should make sure to export this type as well. I'm not sure why the root index.ts in this package enumerates inidividual symbols rather than exporting * from ./components, but you'll want to add this type to that file as well for now

freben replied:
https://github.com/backstage/backstage/pull/8580

freben reviewed (approved):
Approving to unblock? Just some small comments

awanlin commented:
@alexrybch any chance you will be finishing this up? If not I'd take a swing at getting it across the finish line as we'd like to see this feature
